src/ispc.cpp

```diff
@@ -122,9 +129,139 @@ static bool __os_has_avx512_support() {
+// on Windows make comprehensive feature detection unreliable.
+// Example of attempt to detect features across platforms can be found in llvm::sys::getHostCPUFeatures()
+// in llvm/lib/TargetParser/Host.cpp but it doesn't solve issue #3 and it doesn't work on macOS completely.
+// Example of feature detection on Apple can be found in
+// __init_cpu_features_resolver in compiler-rt/lib/builtins/cpu_model/aarch64/fmv/apple.inc.
+
+// To avoid introducing complex and error-prone code into ISPC, let's rely on the LLVM target parser to provide the
+// default features for a particular CPU.
+static std::string lGetARMTargetFeatures(Arch arch, const std::string &cpu) {
+    Assert(arch == Arch::arm || arch == Arch::aarch64);
+    std::string targetFeatures;
+
+    // Keep the features for custom_linux unchanged
+    if (g->target_os == TargetOS::custom_linux) {
+        targetFeatures =
+            (arch == Arch::arm) ? "+crypto,+fp-armv8,+neon,+sha2" : "+aes,+crc,+crypto,+fp-armv8,+neon,+sha2";
+    } else {
+        // Get features for the requested CPU
+        std::vector<llvm::StringRef> features;
+        if (arch == Arch::arm) {
+            llvm::ARM::ArchKind archKind = llvm::ARM::parseCPUArch(cpu);
+            if (archKind == llvm::ARM::ArchKind::INVALID) {
+                Error(SourcePos(), "Invalid CPU name for ARM architecture: %s", cpu.c_str());
+                return "";
+            }
+#if ISPC_LLVM_VERSION >= ISPC_LLVM_17_0
+            using FPUKindType = llvm::ARM::FPUKind;
+#else
+            using FPUKindType = unsigned;
+#endif
+            FPUKindType fpu = llvm::ARM::getDefaultFPU(cpu, archKind);
+            llvm::ARM::getFPUFeatures(fpu, features);
+        } else if (arch == Arch::aarch64) {
+            std::optional<llvm::AArch64::CpuInfo> cpuInfo = llvm::AArch64::parseCpu(cpu);
+            if (!cpuInfo) {
+                Error(SourcePos(), "Invalid CPU name for AArch64 architecture: %s", cpu.c_str());
+                return "";
+            }
+#if ISPC_LLVM_VERSION >= ISPC_LLVM_18_1
+            using CpuExtensionsType = llvm::AArch64::ExtensionBitset;
+#else
+            using CpuExtensionsType = uint64_t;
+#endif
+#if ISPC_LLVM_VERSION >= ISPC_LLVM_17_0
+            CpuExtensionsType cpuExtensions = cpuInfo->getImpliedExtensions();
+#else
+            llvm::AArch64::ArchInfo AI = llvm::AArch64::parseCpu(cpu).Arch;
+            CpuExtensionsType cpuExtensions = getDefaultExtensions(cpu, AI);
+#endif
+            llvm::AArch64::getExtensionFeatures(cpuExtensions, features);
+        } else {
+            UNREACHABLE();
+        }
+        // Sort them for easier testing
+        std::sort(features.begin(), features.end());
+        // Form the features string
+        targetFeatures = llvm::join(features, ",");
+    }
+    return targetFeatures;
+}
+
+// Check if a specific feature is supported
+static bool lIsARMFeatureSupported(const std::string &feature, const std::string &featureString) {
```

Comment:
I don't really like the string-finding logic here. We could have used a vector (initially constructed in lGetARMTargetFeatures) and produced a std::string only when we needed to call m_target->createTargetMachine. What do you think?

Reply:
Yes, it makes sense, I'll change it.